# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [391]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [392]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [393]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [394]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [395]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [396]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [397]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [398]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [399]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [400]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [401]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [402]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [403]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [404]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [405]:
data.duplicated().sum()

71

In [406]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [407]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [408]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [409]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [410]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [411]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

**Цель исследования**:
Необходимо выявить зависимость между количеством детей, семейным положением и доходом на факт погашения кредита в срок. Согласно поставленной задаче выведем следующие гипотезы:

- Количество детей влияет на возврат кредита в срок;
- Семейное положение влияет на возврат кредита в срок;
- Уровень дохода влияет на возврат кредита в срок;
- Цель кредита влияет на возврат кредита в срок.

**Способ исследования**:
Для выявления зависимостей между категорией Клиента и погашением кредита в срок необходимо подсчитать долю наличия задолженности по каждой категории от общего числа Клиентов по заданной категории. Чем выше показатель, тем менее надежным можно считать категорию Клиентов.

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Создадим функцию, которая в качестве аргумента принимает название столбца с категориями и подсчитывает долю наличия задолженности:

In [412]:
# объявляем функцию с помощью ключа def
# для группировки значений по категориям используем метод groupby()
# так как в столбце debt хранятся только значения 0 и 1, что соответствует отсутствию
# и наличию задолженности по кредиту можем использовать метод mean() и умножение его на 100
# используем round() для уменьшения количества знаков после запятой
# и сортируем в порядке убывания методом sort_values()
# для более красивого вывода результата можно использовать метод to_frame()
def category_debt(column):
    share_by_category = data.groupby(column)['debt'].mean() * 100
    return share_by_category.round(2).sort_values(ascending=False).to_frame()

Используем созданную функцию с аргументом 'children' для подсчета доли просрочек в зависимости от количества детей

In [413]:
category_debt('children')

debt
children      
4         9.76
2         9.45
1         9.23
3         8.18
0         7.54
5         0.00

Получив данные можно заметить, что у Клиентов с пятью детьми вовсе нет просрочек по кредиту, что могло быть спровоцировано меньшим количеством в данной категории. Для чистоты ислледований построим сводную таблицу с суммой, количеством и долей просрочек. 

In [414]:
# Группировка таблицы по количеству детей, данные для счета из столбца 'debt', и подсчет суммы,
# количества и доли
category_pivot = pd.pivot_table(data, index=['children'], 
                                values=['debt'], 
                                aggfunc=['sum', 'count', 'mean'])
category_pivot['mean'] *= 100
category_pivot = category_pivot.sort_values(by=('count', 'debt'), ascending=False)
category_pivot.round(2)

sum  count  mean
          debt   debt  debt
children                   
0         1063  14091  7.54
1          444   4808  9.23
2          194   2052  9.45
3           27    330  8.18
4            4     41  9.76
5            0      9  0.00

В полученной таблице количество Клиентов с тремя, четырьмя и пятью детьми значительно ниже, чем у остальных категорий, поэтому эти данные непоказательны в данном исследовании.

Так же можем заметить, что у Клиентов с одним и двумя детьми доля просрочек по кредиту выше почти на 2 пункта, следовательно данные Клиенты менее надежны как заемщики.
При этом существенной разницы доли просрочек между Клиентами с одним или двумя детьми несущественна, что провоцирует создание **новой гипотезы**:
- Наличие детей влияет на возврат кредита в срок.

Для решения новой гипотезы создадим новую категорию Клиентов с категориями "Нет детей" и "Есть дети". В категории "Есть дети" поместим Клиентов с одним или двумя детьми, так как ранее определили, что остальное количество детей не показательно ввиду недостатка данных.

In [415]:
# создаем функцию для категоризации данных
def categorize_children(row):
    if row == 0:
        return 'Нет детей'
    elif row == 1 or row == 2:
        return 'Есть дети'

In [416]:
# создаем столбец с категориями
data['category_children'] = data['children'].apply(categorize_children)

Создаем новую сводную таблицу по новым категориям

In [417]:
# Группировка таблицы по категории наличия детей, данные для счета из столбца 'debt', и подсчет суммы,
# количества и доли
category_pivot_children = pd.pivot_table(data, index=['category_children'], 
                                values=['debt'], 
                                aggfunc=['sum', 'count', 'mean'])
category_pivot_children['mean'] *= 100
category_pivot_children = category_pivot_children.sort_values(by=('count', 'debt'), ascending=False)
category_pivot_children.round(2)

sum  count  mean
                   debt   debt  debt
category_children                   
Нет детей          1063  14091  7.54
Есть дети           638   6860  9.30

Благодаря новой категории четко заметна зависимость возврата кредита в срок от наличия детей.

**Вывод:** 

В итоге проведенного исследования были выявлены следующие факты:

Гипотеза о зависимости возврата кредита в срок от количества детей не может быть подтверждена или опровергнута из-за отсутствия достаточного количества данных о Клиентах с тремя, четырьмя и пятью детьми.

Новая гипотеза о зависимости возврата кредита в срок от наличия детей подтверждена, так как доля прострочек у Клиентов без детей составляет 7.54%, а у Клиентов с детьми для на 1.76% выше - 9.30%

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

В исследовании зависимости возврата кредита в срок от количества детей было выявлено, что наилучшим способом послужит указание суммы, количества и доли просрочек, поэтому в данном исследовании используем его.

In [418]:
# Группировка таблицы по семейному положению, данные для счета из столбца 'debt', и подсчет суммы,
# количества и доли
category_pivot_family = pd.pivot_table(data, index=['family_status'], 
                                values=['debt'], 
                                aggfunc=['sum', 'count', 'mean'])
category_pivot_family['mean'] *= 100
category_pivot_family = category_pivot_family.sort_values(by=('mean', 'debt'), ascending=False)
category_pivot_family.round(2)

sum  count  mean
                      debt   debt  debt
family_status                          
Не женат / не замужем  273   2796  9.76
гражданский брак       385   4134  9.31
женат / замужем        927  12261  7.56
в разводе               84   1189  7.06
вдовец / вдова          63    951  6.62

По полученным данным можно выделить категории, по которым доля просрочек выше:
- Не женат / не замужем
- Гражданский брак

И те категории, по которым доля ниже:
- женат / замужем
- в разводе
- вдовец / вдова

При обобщении категорий с большей и меньшей долей делаем вывод, что Клиенты в официальных отношениях или были в них наиболее надежны чем Клиенты, никогда не состоящие в официальных отношениях.

**Вывод:** 
Гипотеза о зависимости возврата кредита в срок от семейного положения подтверждена. Клиенты, которые состояли или состоят в официальных отношениях чаще возвращают кредит в срок чем Клиенты, которые никогда не состояли в официальных отношениях.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Используем сводную таблицу для получения результатов по категориям уровня дохода.

In [421]:
# Группировка таблицы по уровню дохода, данные для счета из столбца 'debt', и подсчет суммы,
# количества и доли
category_pivot_income = pd.pivot_table(data, index=['total_income_category'], 
                                values=['debt'], 
                                aggfunc=['sum', 'count', 'mean'])
category_pivot_income['mean'] *= 100
category_pivot_income = category_pivot_income.sort_values(by=('mean', 'debt'), ascending=False)
category_pivot_income.round(2)

sum  count  mean
                       debt   debt  debt
total_income_category                   
E                         2     22  9.09
C                      1353  15921  8.50
A                         2     25  8.00
B                       354   5014  7.06
D                        21    349  6.02

По полученным данным видим, что Клиентов в категориях 'D'(30001- 50000), 'A'(более 1000000) и 'E'(до 30000) мало для проведения полноценного исследования, их доля просрочек является непоказательной, а двух категорий 'C'(50001-200000) и 'B'(200001-1000000) мало для поиска зависимости возврата кредита в срок от уровня дохода.

**Вывод:** 
Нет возможности подтвердить или опровергнуть гипотезу из-за недостатка данных о Клиентах с категориями уровня дохода 'D', 'A' и 'E'

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Используем сводную таблицу для получения результатов по категориям цели кредита.

In [420]:
# Группировка таблицы по цели кредита, данные для счета из столбца 'debt', и подсчет суммы,
# количества и доли
category_pivot_purpose = pd.pivot_table(data, index=['purpose_category'], 
                                values=['debt'], 
                                aggfunc=['sum', 'count', 'mean'])
category_pivot_purpose['mean'] *= 100
category_pivot_purpose = category_pivot_purpose.sort_values(by=('mean', 'debt'), ascending=False)
category_pivot_purpose.round(2)

sum  count  mean
                         debt   debt  debt
purpose_category                          
операции с автомобилем    400   4279  9.35
получение образования     369   3988  9.25
проведение свадьбы        183   2313  7.91
операции с недвижимостью  780  10751  7.26

**Вывод:** 
Клиенты с целью кредитования для операций с недвижимостью и проведение свадьбы ниже, чем у клиентов с целью кредитования для операций с автомобилем и получения образования.

Гипотезу о зависимости возврата кредита в срок от цели кредита можно подтвердить.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Причинами могут стать ошибки системы или фактическое отсутствие данных о Клиенте. 
В нашей таблице пропуски находятся в столбцах c трудовым стажем и ежемесячным доходом, причем в равном количестве - 2174 строки.

Так же пропуски в этих столбцах есть только у Клиентов с типом занятости пенсионер и госслужащий.
У пенсионеров и госслужащих явно есть ежемесячный доход - пенсия и зарплата соответственно, как и должен иметься трудовой стаж. Поэтому сделаю вывод, что пропуски спровоцированы ошибкой выгрузки данных по Клиентам с типом занятости пенсионер и госслужащий.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Данные могут быть неравномерны, со значениями в несколько раз превышающими максимальное из других значений, что в свою очередь существенно искажает среднее значение. При этом медианное значение не подвержено таким искажениям. 

### Шаг 4: общий вывод.

**Задача исследования**:
Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Исходя из задачи поставлены следующие гипотезы:
- Гипотеза 1: Количество детей влияет на возврат кредита в срок;
- Гипотеза 2: Семейное положение влияет на возврат кредита в срок;
- Гипотеза 3: Уровень дохода влияет на возврат кредита в срок;
- Гипотеза 4: Цель кредита влияет на возврат кредита в срок.

**Гипотеза 1**

В ходе исследования был выявлен недостаток данных по Клиентам с 3, 4 и 5 детьми
Доли просрочек по кредиту для категорий Клиентов:
- Клиенты без детей - *7.54%*
- Клиенты с одним ребенком - *9.23%*
- Клиенты с двумя детьми - *9.45%*
За недостатком данных можно выдвинуть новую гипотезу:
- Наличие детей влияет на возврат кредита в срок.
Для данной гипотезы сформированы следующие данные по доле просрочек по кредиту:
- Клиенты без детей - *7.54%
- Клиенты с детьми - *9.30%*
Заметно влияние наличия детей на возврат кредита в срок, доля Клиентов с детьми имеющие просрочку по кредиту на 1.76% больше Клиентов без детей.

*Вывод*: Необходимо провести исследование на большей выборке данных. До проведения исследования можно использовать категории по наличию/отсутствию детей для модели кредитного скоринга.

**Гипотеза 2**

Доли просрочек по кредиту для категорий Клиентов:
- Не женат / не замужем	- *9.76%*
- гражданский брак - *9.31%*
- женат / замужем - *7.56%*
- в разводе - *7.06%*
- вдовец / вдова - *6.62%*

При обобщении категорий с большей и меньшей долями делаем вывод, что Клиенты в официальных отношениях или были в них наиболее надежны чем Клиенты, никогда не состоящие в официальных отношениях.

*Вывод*: Гипотеза подтверждена. Категории Клиентов по семейному положению можно использовать для модели кредитного скоринга. 

**Гипотеза 3**

В ходе исследования был выявлен недостаток данных по Клиентам с категориями дохода 'A'(1000001 и выше), 'D'(30001–50000) и 'E'(0 - 30000)
Доли просрочек по кредиту для остальных категорий Клиентов:
- B(200001–1000000) - *7.06%*
- C(50001–200000) - *8.50%*
Нет возможности подтвердить или опровергнуть гипотезу из-за недостатка данных о Клиентах с категориями уровня дохода 'D', 'A' и 'E'

*Вывод*: Необходимо провести исследование на большей выборке данных.

**Гипотеза 4**

Доли просрочек по кредиту для категорий Клиентов:
- операции с автомобилем - *9.35%*
- получение образования - *9.25%*
- проведение свадьбы - *7.91%*
- операции с недвижимостью - *7.26%*
Клиенты с целью кредитования для операций с недвижимостью и проведение свадьбы ниже чем у клиентов с целью кредитования для операций с автомобилем и получения образования.

*Вывод*: Гипотеза подтверждена. Категории Клиентов по цели кредита можно использовать для модели кредитного скоринга.

**Рекомендации**:
1. Провести дополнительное исследование зависимости возврата кредита в срок от количества детей на большей выборке данных. 
2. Провести дополнительное исследование зависимости возврата кредита в срок от уровня дохода на большей выборке данных. 
3. Использовать для модели кредитного скоринга зависимости от семейного положения и от цели кредита.
4. При необходимости и до дополнительного исследования можно использовать зависимость от наличия детей для модели кредитного скоринга.
